### DISCUSSION TITLE
Add security headers as middlewares

### DISCUSSION THREAD

**Discussion by Niccolum** [fastapi/fastapi#8548] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
...
```

### Description

Add couple of security middleware, which, I think, must be built-in app

### Wanted Solution

Implement middlewares from https://github.com/tmotagam/Secweb :
Required (my opinion):
- Content Security Policy (CSP)
- Referrer Policy
- HTTP Strict Transport Security(HSTS)
- X-XSS-Protection

Other - as you want (my opinion)
- ExpectCT
- Origin Agent Cluster
- X-Content-Type-Options
- X-DNS-Prefetch-Control
- X-Download-Options
- X-Frame
- X-Permitted-Cross-Domain-Policies
- Cross-Origin-Embedder-Policy
- Cross-Origin-Opener-Policy
- Cross-Origin-Resource-Policy

### Wanted Code

```python
from fastapi import FastAPI
from fastapi.middleware import ContentSecurityPolicy # or other middleware

app = FastAPI()

app.add_middleware(ContentSecurityPolicy, Option=...)
```

### Alternatives

https://github.com/tmotagam/Secweb

### Operating System

Linux, Windows, macOS

### FastAPI Version

latest

### Python Version

3.10.1

**Comment by lucastosetto** [comment]
@Niccolum That's a great suggestion and I can work on this enhancement, but I think it would be more appropriate to do it on [starlette](https://github.com/encode/starlette) and then import the packages here, as it's already implemented nowadays. For instance:

`cors.py`

`from starlette.middleware.cors import CORSMiddleware as CORSMiddleware  # noqa`

**Comment by Niccolum** [comment]
@lucastosetto Starlette maintainer said, that he see this only as 3rd party library.

**Comment by Cielquan** [comment]
I came across this issue while searching something else, but I implemented such a middleware in a proprietary project.

You may or may not take this code as a starting point if you wish:
```py
"""Middleware for security."""
from collections import OrderedDict

from fastapi import FastAPI, Request, Response
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint


CSP: dict[str, str | list[str]] = {
    "default-src": "'self'",
    "img-src": [
        "*",
        # For SWAGGER UI
        "data:",
    ],
    "connect-src": "'self'",
    "script-src": "'self'",
    "style-src": ["'self'", "'unsafe-inline'"],
    "script-src-elem": [
        # For SWAGGER UI
        "https://cdn.jsdelivr.net/npm/swagger-ui-dist@4/swagger-ui-bundle.js",
        "'sha256-1I8qOd6RIfaPInCv8Ivv4j+J0C6d7I8+th40S5U/TVc='",
    ],
    "style-src-elem": [
        # For SWAGGER UI
        "https://cdn.jsdelivr.net/npm/swagger-ui-dist@4/swagger-ui.css",
    ],
}


def parse_policy(policy: dict[str, str | list[str]] | str) -> str:
    """Parse a given policy dict to string."""
    if isinstance(policy, str):
        # parse the string into a policy dict
        policy_string = policy
        policy = OrderedDict()

        for policy_part in policy_string.split(";"):
            policy_parts = policy_part.strip().split(" ")
            policy[policy_parts[0]] = " ".join(policy_parts[1:])

    policies = []
    for section, content in policy.items():
        if not isinstance(content, str):
            content = " ".join(content)
        policy_part = f"{section} {content}"

        policies.append(policy_part)

    parsed_policy = "; ".join(policies)

    return parsed_policy


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """Add security headers to all responses."""

    def __init__(self, app: FastAPI, csp: bool = True) -> None:
        """Init SecurityHeadersMiddleware.

        :param app: FastAPI instance
        :param no_csp: If no CSP should be used;
            defaults to :py:obj:`False`
        """
        super().__init__(app)
        self.csp = csp

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        """Dispatch of the middleware.

        :param request: Incoming request
        :param call_next: Function to process the request
        :return: Return response coming from from processed request
        """
        headers = {
            "Content-Security-Policy": "" if not self.csp else parse_policy(CSP),
            "Cross-Origin-Opener-Policy": "same-origin",
            "Referrer-Policy": "strict-origin-when-cross-origin",
            "Strict-Transport-Security": "max-age=31556926; includeSubDomains",
            "X-Content-Type-Options": "nosniff",
            "X-Frame-Options": "DENY",
            "X-XSS-Protection": "1; mode=block",
        }
        response = await call_next(request)
        response.headers.update(headers)

        return response

```

You can add the middleware the default way like:
```py
app = FastAPI()

app.add_middleware(SecurityHeadersMiddleware, csp=True)
```

**This code needs python 3.10+ because of the type hints only.**

**Comment by mert-kurttutan** [comment]
What is the current status of this issue? Is there a plan to add such feature?
If there is no, I guess there are two options to implement such security check: Use the alternative [https://github.com/tmotagam/Secweb](https://github.com/tmotagam/Secweb) (as suggested) or implement it yourself

Any comment is appreciated !!

**Comment by staubina** [comment]
It sounds like many are running into the same issue, what did you end up using @mert-kurttutan and @Niccolum?

**Comment by Niccolum** [comment]
@staubina i used library secweb, but I don't really like its implementation

**Comment by karolpawlowski** [comment]
What do you guys think about this library https://secure.readthedocs.io/en/latest/index.html as an alternative to the proposed one https://github.com/tmotagam/Secweb?

**Reply by EverStarck** [reply]
I really liked this one, but seems a bit abandoned, I submitted an issue and didn't get a response in 3 weeks, also there's another issues 3months ago without answer :/

**Comment by stevleibelt** [comment]
@tiangolo I've read the [previous issue](https://github.com/tiangolo/fastapi/issues/4420) and this discussion while doing a research on finding the best and most common approach for "implement security headers".

Is there anything fastapi itself would recommend? I could almost find the right spot in the [documentation](https://fastapi.tiangolo.com/advanced/middleware/).
From my side, a starting point would be good or even a "this is the recommended way" would be even better - or as good as a "think about this points first" section.

Right now it looks like a lot of open question.

Best regards and thanks in advance for any answer :-).

**Comment by inanpy** [comment]
Hi everyone 👋

I recently built a package called [fastapi-armor](https://github.com/inanpy/fastapi-armor) to address this very need, similar in spirit to **Django’s SecurityMiddleware** or **helmet.js from the Node.js** ecosystem.

It’s a FastAPI middleware that sets common security-related HTTP headers, such as Content-Security-Policy, Strict-Transport-Security, X-Frame-Options, and more.

The integration is minimal, just one line to add the middleware, and it includes useful presets (strict, relaxed, none) for different use cases.

Headers can also be customized individually as needed.

📦 Available on PyPI: https://pypi.org/project/fastapi-armor/

Hope it helps others looking for a simple and effective solution to implement security headers in FastAPI.

I’m open to feedback and happy to collaborate if this could serve as a base for something more official in the future.
